src/event-bus/EventStore.ts

```diff
@@ -0,0 +1,130 @@
+import { Event, PersistedEvent } from '../shared/events';
+import { Logger } from '../shared/logger';
+import { promisify } from '../shared/promisify';
+import { sleep } from '../shared/sleep';
+
+const eventStoreConstructor = require('eventstore');
+
+export interface EventStoreEvent<T> {
+  streamId: string;
+  aggregateId: string;
+  aggregate: any;
+  context: any;
+  streamRevision: number;
+  commitId: string;
+  commitSequence: number;
+  commitStamp: Date;
+  payload: T;
+  id: string;
+  restInCommitStream: number;
+}
+
+export interface EventStoreStream<T> {
+  streamId: string;
+  aggregateId: string;
+  aggregate: any;
+  context: any;
+  events: EventStoreEvent<T>[];
+  id: string;
+  lastRevision: number;
+}
+
+export default class EventStore {
+
+  private isConnected = false;
+  private eventStore: any;
+
+  constructor(
+    private readonly publisher: (event: Event<any>, cb: (err: any) => void) => void,
+    private readonly eventStoreConfig: any,
+    private readonly logger: Logger,
+    private readonly numRetries = 1,
+    private readonly sleepFor = 300,
+  ) { }
+
+  public async persistEvent(event: Event<any>) {
+    for (let i = 0; i < this.numRetries + 1; i++) {
+      const persisted = await this.tryPersistEvent(event);
+      if (persisted === true) {
+        return true;
+      }
+      if (i < this.numRetries) {
+        await sleep(this.sleepFor);
+      }
+    }
+    return false;
+  }
+
+  private async tryPersistEvent(event: Event<any>) {
+    try {
+      await this.connect();
+      const stream = await this.eventStore.getLastEventAsStream({
+        aggregateId: String(event.teamId),
+      });
+      stream.addEvent(event);
+      await this.eventStore.commit(stream);
+      return true;
+    } catch (err) {
+      this.logger.error('Unable to post event %s', event.type, err);
+      this.isConnected = false;
+      return false;
+    }
+  }
+
+  public async getEvents(teamId: number, since: number = 0): Promise<PersistedEvent<any>[] | false> {
+    for (let i = 0; i < this.numRetries + 1; i++) {
+      const stream = await this.tryGetStream(teamId, since);
+      if (stream) {
+        return stream.map((event: any, j: number) => {
+          event.payload.streamRevision = since + j;
```

Comment:
This code is unchanged in the current PR. The streamRevision is overwritten since I remember there were some inconsistencies with it and we know what it should be anyway. I'd prefer to address the rest of the questions in a separate PR if deemed necessary.